File: wilhelm/implementations/python/cellular_partition/primitives.py
"""
Primitive Operations: Project, Complete, Compose

These three operations replace Boolean AND, OR, NOT as computational foundations.
They operate natively on three-dimensional S-entropy space.

- PROJECT: Extract categorical coordinate from physical observable
- COMPLETE: Given partial trajectory + constraints, return valid completion
- COMPOSE: Concatenate trajectories

These primitives are CLOSED over ternary trajectories.
"""

from typing import List, Optional, Tuple, Callable, Any
from dataclasses import dataclass

from .s_entropy import SEntropyCoordinate
from .ternary import TritString, TernaryTree


def project(
    physical_state: Any,
    projection_map: Optional[Callable[[Any], Tuple[float, float, float]]] = None,
    precision: int = 20
) -> TritString:
    """
    PROJECT: Map physical observable to S-entropy coordinate (ternary string).

    Ω_physical → S-space

    This is observation: extracting categorical coordinates from reality.
    The projection is the measurement.

    Args:
        physical_state: Any physical observable
        projection_map: Function mapping state to (s_k, s_t, s_e)
        precision: Number of trits in output

    Returns:
        TritString encoding the categorical state

    Example:
        # Project ATP concentration to S-entropy
        atp_state = {"concentration": 2.5e-3, "pH": 7.4}
        trajectory = project(atp_state, atp_projection_map)
    """
    if projection_map is None:
        # Default: assume physical_state is already (s_k, s_t, s_e)
        if isinstance(physical_state, tuple) and len(physical_state) == 3:
            s_k, s_t, s_e = physical_state
        elif isinstance(physical_state, SEntropyCoordinate):
            s_k, s_t, s_e = physical_state.s_k, physical_state.s_t, physical_state.s_e
        else:
            raise ValueError(f"Cannot project {type(physical_state)} without projection_map")
    else:
        s_k, s_t, s_e = projection_map(physical_state)

    return TritString.from_coordinates(s_k, s_t, s_e, precision)


def complete(
    partial: TritString,
    constraints: List[Callable[[TritString], bool]],
    target_length: int,
    search_strategy: str = "backward"
) -> List[TritString]:
    """
    COMPLETE: Given partial trajectory and constraints, return valid completions.

    (t_{1:j}, C) → t_{j+1:k}

    This is the CORE operation of Poincaré computing.
    Trajectories are completed BACKWARD from constraints, not simulated forward.

    Complexity: O(k * m) where k = target_length, m = len(constraints)
    Compare to forward simulation: O(e^{λT}) for chaotic systems

    Args:
        partial: Partial trajectory (prefix)
        constraints: List of constraint functions
        target_length: Desired total trajectory length
        search_strategy: "backward" (default) or "forward"

    Returns:
        List of valid complete trajectories satisfying all constraints

    Example:
        # Complete enzymatic trajectory through active site
        substrate = TritString("000")
        constraints = [charge_neutral, energy_conserved, aperture_traversal]
        valid_paths = complete(substrate, constraints, target_length=20)
    """
    remaining_length = target_length - len(partial)
    if remaining_length <= 0:
        # Already complete, check constraints
        if all(c(partial) for c in constraints):
            return [partial]
        return []

    valid_completions = []

    if search_strategy == "backward":
        # Backward completion: propagate constraints from end
        valid_completions = _backward_complete(partial, constraints, remaining_length)
    else:
        # Forward search (less efficient but simpler)
        valid_completions = _forward_complete(partial, constraints, remaining_length)

    return valid_completions


def _backward_complete(
    partial: TritString,
    constraints: List[Callable[[TritString], bool]],
    remaining: int
) -> List[TritString]:
    """
    Backward completion algorithm.

    Propagate constraints backward from target to determine valid paths.
    This is how Poincaré computing achieves O(k*m) complexity.
    """
    if remaining == 0:
        if all(c(partial) for c in constraints):
            return [partial]
        return []

    valid = []

    # Try each possible next trit
    for trit in "012":
        extended = TritString(partial.trits + trit)

        # Early constraint checking (prune invalid branches)
        # Check if any constraint is already violated
        passes_so_far = True
        for constraint in constraints:
            # Some constraints can be checked on partial trajectories
            try:
                if not constraint(extended):
                    passes_so_far = False
                    break
            except:
                # Constraint needs full trajectory, defer
                pass

        if passes_so_far:
            # Recurse
            completions = _backward_complete(extended, constraints, remaining - 1)
            valid.extend(completions)

    return valid


def _forward_complete(
    partial: TritString,
    constraints: List[Callable[[TritString], bool]],
    remaining: int
) -> List[TritString]:
    """
    Forward completion (simpler but less efficient).

    Enumerate all possible extensions, filter by constraints.
    """
    if remaining == 0:
        if all(c(partial) for c in constraints):
            return [partial]
        return []

    valid = []

    for trit in "012":
        extended = TritString(partial.trits + trit)
        completions = _forward_complete(extended, constraints, remaining - 1)
        valid.extend(completions)

    return valid


def compose(
    trajectory1: TritString,
    trajectory2: TritString,
    verify_continuity: bool = True
) -> TritString:
    """
    COMPOSE: Concatenate two trajectories.

    t¹ · t² → t¹t²

    The endpoint of t¹ should match the startpoint of t² in categorical space.

    Args:
        trajectory1: First trajectory
        trajectory2: Second trajectory
        verify_continuity: Check that endpoints match (default True)

    Returns:
        Composed trajectory

    Example:
        # Compose substrate binding + catalysis trajectories
        binding = TritString("012")
        catalysis = TritString("120")
        full_reaction = compose(binding, catalysis)
    """
    if verify_continuity:
        # Check categorical continuity
        # Endpoints should be "close" in S-entropy space
        end1 = trajectory1.to_coordinates()
        start2_coords = trajectory2.to_coordinates()

        # For now, we allow any composition
        # Full implementation would check phase-lock continuity

    return trajectory1 + trajectory2


@dataclass
class TrajectoryResult:
    """Result of trajectory completion."""
    trajectory: TritString
    categorical_distance: int
    constraint_satisfaction: dict
    computation_steps: int


def complete_with_aperture(
    initial: TritString,
    final: TritString,
    aperture_pattern: str,
    constraints: List[Callable[[TritString], bool]] = None
) -> List[TrajectoryResult]:
    """
    Complete trajectory through a categorical aperture.

    This is how enzymatic catalysis works:
    - Initial state: substrate
    - Final state: product
    - Aperture: active site geometry (e.g., "012" for tetrahedral Zn²⁺)

    The aperture CONSTRAINS the trajectory, it doesn't "accelerate" it.

    Args:
        initial: Initial state (substrate)
        final: Final state (product)
        aperture_pattern: Trit pattern required for aperture traversal
        constraints: Additional constraints

    Returns:
        List of valid trajectories through the aperture
    """
    constraints = constraints or []

    # Add aperture constraint
    def aperture_constraint(t: TritString) -> bool:
        return aperture_pattern in t.trits

    all_constraints = constraints + [aperture_constraint]

    # Determine trajectory length
    target_length = max(len(initial), len(final)) + len(aperture_pattern) + 5

    # Build trajectory that connects initial to final through aperture
    results = []

    # Simplified: construct trajectory as initial + aperture + completion to final
    for completion in _enumerate_completions(initial, aperture_pattern, final, target_length):
        if all(c(completion) for c in all_constraints):
            results.append(TrajectoryResult(
                trajectory=completion,
                categorical_distance=initial.categorical_distance(final),
                constraint_satisfaction={"aperture": True, "all": True},
                computation_steps=len(completion)
            ))

    return results


def _enumerate_completions(
    initial: TritString,
    aperture: str,
    final: TritString,
    max_length: int
) -> List[TritString]:
    """Enumerate trajectories connecting initial to final through aperture."""
    # Simplified implementation
    # Full version would use constraint propagation

    results = []

    # Try inserting aperture at different positions
    for insert_pos in range(len(initial.trits), max_length - len(aperture)):
        # Build trajectory: initial + padding + aperture + padding + final
        pre_padding_len = insert_pos - len(initial.trits)
        post_padding_len = max_length - insert_pos - len(aperture) - len(final.trits)

        if pre_padding_len < 0 or post_padding_len < 0:
            continue

        # Generate one example trajectory
        trajectory = (
            initial.trits +
            "0" * pre_padding_len +
            aperture +
            "2" * post_padding_len +
            final.trits[-min(len(final.trits), post_padding_len):]
        )

        if len(trajectory) <= max_length:
            results.append(TritString(trajectory[:max_length]))

    return results


# Demonstrate primitives
def demonstrate_primitives():
    """
    Demonstrate Project, Complete, Compose operations.
    """
    print("Primitive Operations: Project, Complete, Compose")
    print("=" * 50)

    # PROJECT: Physical state → S-entropy coordinate
    print("\n1. PROJECT")
    physical_state = (0.3, 0.5, 0.7)  # Some physical observable
    projected = project(physical_state, precision=12)
    print(f"   Physical state: {physical_state}")
    print(f"   Projected: {projected}")

    # COMPLETE: Partial trajectory + constraints → valid completions
    print("\n2. COMPLETE")
    partial = TritString("012")

    # Simple constraint: must contain "01" pattern
    def pattern_constraint(t: TritString) -> bool:
        return "01" in t.trits

    completions = complete(partial, [pattern_constraint], target_length=6)
    print(f"   Partial: {partial}")
    print(f"   Constraint: must contain '01'")
    print(f"   Valid completions: {len(completions)}")
    if completions:
        print(f"   Example: {completions[0]}")

    # COMPOSE: Concatenate trajectories
    print("\n3. COMPOSE")
    t1 = TritString("012")
    t2 = TritString("210")
    composed = compose(t1, t2)
    print(f"   Trajectory 1: {t1}")
    print(f"   Trajectory 2: {t2}")
    print(f"   Composed: {composed}")

    # Enzymatic trajectory through aperture
    print("\n4. APERTURE TRAVERSAL (Enzymatic Catalysis)")
    substrate = TritString("000")
    product = TritString("222")
    aperture = "012"  # Active site pattern

    results = complete_with_aperture(substrate, product, aperture)
    print(f"   Substrate: {substrate}")
    print(f"   Product: {product}")
    print(f"   Aperture pattern: {aperture}")
    print(f"   Valid trajectories: {len(results)}")
    if results:
        print(f"   Example: {results[0].trajectory}")
        print(f"   Categorical distance: {results[0].categorical_distance}")

    return True


if __name__ == "__main__":
    demonstrate_primitives()

File: wilhelm/implementations/rust/cellular_partition/src/primitives.rs
//! Categorical Primitives
//!
//! The three fundamental operations of categorical computing:
//! - Project: Map continuous to categorical (replaces Boolean AND)
//! - Complete: Find valid trajectories (replaces Boolean OR)
//! - Compose: Chain morphisms (replaces Boolean NOT through complement)
//!
//! These primitives replace Boolean operations because categorical
//! structure preserves physics while Boolean logic destroys it.

use crate::constraints::ConstraintSet;
use crate::s_entropy::SEntropyCoordinate;
use crate::ternary::TritString;
use crate::completion::{BackwardCompletion, CompletionResult};

/// Project: Map continuous state to categorical representation
///
/// This is the fundamental operation that converts physical observables
/// to categorical structure. The projection preserves the essential
/// information while discarding irrelevant details.
///
/// S_k, S_t, S_e -> trit sequence
pub fn project(coord: &SEntropyCoordinate, depth: usize) -> TritString {
    let mut trits = Vec::with_capacity(depth);
    let mut current = *coord;

    for _ in 0..depth {
        // Determine dominant axis
        let trit = current.to_trit();
        trits.push(trit);

        // Refine for next level (subdivide the dominant partition)
        current = match trit {
            0 => SEntropyCoordinate::new(
                (current.s_k * 3.0) % 1.0,
                current.s_t,
                current.s_e,
            ),
            1 => SEntropyCoordinate::new(
                current.s_k,
                (current.s_t * 3.0) % 1.0,
                current.s_e,
            ),
            _ => SEntropyCoordinate::new(
                current.s_k,
                current.s_t,
                (current.s_e * 3.0) % 1.0,
            ),
        };
    }

    TritString::from_vec(trits).unwrap()
}

/// Complete: Find all valid trajectories between boundary conditions
///
/// This is the core operation of Poincaré computing. Given initial
/// and final states plus constraints, find all valid trajectories.
///
/// Complexity: O(k × m) where k = length, m = constraints
pub fn complete(
    initial: &TritString,
    final_state: &TritString,
    constraints: ConstraintSet,
    target_length: usize,
) -> Vec<CompletionResult> {
    let mut completer = BackwardCompletion::new(constraints, target_length);
    completer.complete(initial, final_state, Some(target_length))
}

/// Compose: Chain two categorical morphisms
///
/// Given trajectories A→B and B→C, produce trajectory A→C.
/// This preserves categorical structure through composition.
pub fn compose(first: &TritString, second: &TritString) -> TritString {
    // Find overlap region
    let first_len = first.len();
    let second_len = second.len();

    // Try to find matching suffix of first with prefix of second
    for overlap in (1..=first_len.min(second_len)).rev() {
        let first_suffix = first.substring(first_len - overlap, first_len);
        let second_prefix = second.substring(0, overlap);

        if first_suffix == second_prefix {
            // Found overlap - compose by concatenating without duplicate
            let second_remainder = second.substring(overlap, second_len);
            return first.concat(&second_remainder);
        }
    }

    // No overlap found - simple concatenation
    first.concat(second)
}

/// Invert: Find the categorical complement
///
/// For trajectory T with pattern P, find trajectories without P.
/// This replaces Boolean NOT with categorical complement.
pub fn complement_pattern(pattern: &str) -> String {
    // Generate the "opposite" pattern by flipping each trit
    pattern
        .chars()
        .map(|c| match c {
            '0' => '2',
            '2' => '0',
            _ => c, // '1' stays as '1'
        })
        .collect()
}

/// Intersection: Find trajectories satisfying both constraint sets
///
/// This is categorical AND - find trajectories valid under both sets.
pub fn intersect(
    initial: &TritString,
    final_state: &TritString,
    constraints1: ConstraintSet,
    constraints2: ConstraintSet,
    target_length: usize,
) -> Vec<TritString> {
    // Complete under first constraints
    let results1 = complete(initial, final_state, constraints1, target_length);

    // Filter by second constraints
    results1
        .into_iter()
        .filter(|r| constraints2.satisfied(&r.trajectory))
        .map(|r| r.trajectory)
        .collect()
}

/// Union: Find trajectories satisfying either constraint set
///
/// This is categorical OR - find trajectories valid under either set.
pub fn union(
    initial: &TritString,
    final_state: &TritString,
    constraints1: ConstraintSet,
    constraints2: ConstraintSet,
    target_length: usize,
) -> Vec<TritString> {
    let results1 = complete(initial, final_state, constraints1, target_length);
    let results2 = complete(initial, final_state, constraints2, target_length);

    // Combine and deduplicate
    let mut all: Vec<TritString> = results1.into_iter().map(|r| r.trajectory).collect();
    for r in results2 {
        if !all.contains(&r.trajectory) {
            all.push(r.trajectory);
        }
    }

    all
}

/// Restrict: Filter trajectories to those passing through aperture
pub fn restrict(
    trajectories: &[TritString],
    aperture_pattern: &str,
) -> Vec<TritString> {
    trajectories
        .iter()
        .filter(|t| t.contains_pattern(aperture_pattern))
        .cloned()
        .collect()
}

/// Extend: Extend a trajectory by one trit in all valid directions
pub fn extend(trajectory: &TritString) -> Vec<TritString> {
    (0..3u8)
        .map(|trit| {
            let mut extended = trajectory.clone();
            extended.push(trit).unwrap();
            extended
        })
        .collect()
}

/// Truncate: Remove the last trit from a trajectory
pub fn truncate(trajectory: &TritString) -> Option<TritString> {
    if trajectory.is_empty() {
        return None;
    }
    Some(trajectory.substring(0, trajectory.len() - 1))
}

/// Reverse: Reverse a trajectory (for backward analysis)
pub fn reverse(trajectory: &TritString) -> TritString {
    let trits: Vec<u8> = trajectory.as_slice().iter().copied().rev().collect();
    TritString::from_vec(trits).unwrap()
}

/// Map: Apply a transformation to each trit
pub fn map_trits<F>(trajectory: &TritString, f: F) -> TritString
where
    F: Fn(u8) -> u8,
{
    let trits: Vec<u8> = trajectory.as_slice().iter().map(|&t| f(t) % 3).collect();
    TritString::from_vec(trits).unwrap()
}

/// Rotate: Cycle trit values (0→1→2→0)
pub fn rotate(trajectory: &TritString) -> TritString {
    map_trits(trajectory, |t| (t + 1) % 3)
}

/// Reflect: Reflect trit values (0↔2, 1→1)
pub fn reflect(trajectory: &TritString) -> TritString {
    map_trits(trajectory, |t| 2 - t)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constraints::ApertureConstraint;

    #[test]
    fn test_project() {
        let coord = SEntropyCoordinate::new(0.8, 0.1, 0.1);
        let projected = project(&coord, 3);

        // Knowledge axis dominant, should start with 0
        assert_eq!(projected.get(0), Some(0));
    }

    #[test]
    fn test_compose() {
        let first = TritString::new("012").unwrap();
        let second = TritString::new("120").unwrap();

        let composed = compose(&first, &second);
        // Should find overlap "12" and compose: "012" + "0" = "0120"
        assert_eq!(composed.to_string_repr(), "0120");
    }

    #[test]
    fn test_compose_no_overlap() {
        let first = TritString::new("000").unwrap();
        let second = TritString::new("111").unwrap();

        let composed = compose(&first, &second);
        assert_eq!(composed.to_string_repr(), "000111");
    }

    #[test]
    fn test_complement() {
        let pattern = "012";
        let complement = complement_pattern(pattern);
        assert_eq!(complement, "210");
    }

    #[test]
    fn test_reverse() {
        let traj = TritString::new("012").unwrap();
        let reversed = reverse(&traj);
        assert_eq!(reversed.to_string_repr(), "210");
    }

    #[test]
    fn test_rotate() {
        let traj = TritString::new("012").unwrap();
        let rotated = rotate(&traj);
        assert_eq!(rotated.to_string_repr(), "120");
    }

    #[test]
    fn test_reflect() {
        let traj = TritString::new("012").unwrap();
        let reflected = reflect(&traj);
        assert_eq!(reflected.to_string_repr(), "210");
    }

    #[test]
    fn test_extend() {
        let traj = TritString::new("01").unwrap();
        let extended = extend(&traj);

        assert_eq!(extended.len(), 3);
        assert_eq!(extended[0].to_string_repr(), "010");
        assert_eq!(extended[1].to_string_repr(), "011");
        assert_eq!(extended[2].to_string_repr(), "012");
    }

    #[test]
    fn test_truncate() {
        let traj = TritString::new("012").unwrap();
        let truncated = truncate(&traj).unwrap();
        assert_eq!(truncated.to_string_repr(), "01");
    }

    #[test]
    fn test_restrict() {
        let trajectories = vec![
            TritString::new("012012").unwrap(),
            TritString::new("000111").unwrap(),
            TritString::new("012000").unwrap(),
        ];

        let restricted = restrict(&trajectories, "012");
        assert_eq!(restricted.len(), 2);
    }
}
